pkg/applicationserver/io/packages/loradms/v1/package.go

```diff
@@ -0,0 +1,132 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loraclouddevicemanagementv1
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"go.thethings.network/lorawan-stack/pkg/applicationserver/io"
+	"go.thethings.network/lorawan-stack/pkg/applicationserver/io/packages"
+	"go.thethings.network/lorawan-stack/pkg/applicationserver/io/packages/loradms/v1/api"
+	"go.thethings.network/lorawan-stack/pkg/applicationserver/io/packages/loradms/v1/api/objects"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"google.golang.org/grpc"
+)
+
+// DeviceManagementPackage is the LoRa Cloud Device Management application package.
+type DeviceManagementPackage struct {
+	server   io.Server
+	registry packages.Registry
+}
+
+// RegisterServices implements packages.ApplicationPackageHandler.
+func (p *DeviceManagementPackage) RegisterServices(s *grpc.Server) {}
+
+// RegisterHandlers implements packages.ApplicationPackageHandler.
+func (p *DeviceManagementPackage) RegisterHandlers(s *runtime.ServeMux, conn *grpc.ClientConn) {}
+
+var (
+	errDeviceEUIMissing    = errors.DefineNotFound("device_eui_missing", "device EUI `{devEUI}` not found")
+	errUplinkRequestFailed = errors.DefineInternal("uplink_request_failed", "uplink request failed")
+)
+
+// HandleUp implements packages.ApplicationPackageHandler.
+func (p *DeviceManagementPackage) HandleUp(ctx context.Context, assoc *ttnpb.ApplicationPackageAssociation, up *ttnpb.ApplicationUp) error {
+	ctx = log.NewContextWithField(ctx, "namespace", "applicationserver/io/packages/loraclouddevicemanagementv1")
+	logger := log.FromContext(ctx)
+
+	eui := objects.EUI(*up.DevEUI)
```

Comment:
I believe that `DevEUI` is not mandatory, i.e. for ABP devices. I think we must skip if we don't have the EUI.

Reply:
The EUI is mandatory as part of the API specification. See [here](https://www.loracloud.com/documentation/device_management?url=v1.html#deliver-uplink-message) (the EUI is marked as required).

Reply:
Yes, but not in our API, so `(*ttnpb.ApplicationUp).DevEUI` may panic

Reply:
Please gracefully log and stop processing here if the end device's `DevEUI` is not set.